Migrations 3.0

goetas opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes
| Fixed issues | https://github.com/doctrine/migrations/issues/857 https://github.com/doctrine/migrations/issues/853 https://github.com/doctrine/migrations/issues/726 https://github.com/doctrine/migrations/issues/805 https://github.com/doctrine/migrations/issues/742 https://github.com/doctrine/migrations/pull/854

#### Summary
This PR proposes what could become doctrine/migrations 3.0. (Some code and ideas are borrowed from https://github.com/doctrine/migrations/pull/162)

This is a "work in progress", so any feedback is appreciated. 

Exiting user-facing features are have been kept as they are ( some BC breaks are present of course).

Summary of changes:
-  The "version" is not anymore a number or date, but the FQCN of the migration class.
-  Multiple namespaces are allowed (and paths). This allows to provide third party migrations, plugin-provided migrations and and dependent migrations as well (as described in  #853)
- Extracted the "metadata table"  logic in its own interface as in https://github.com/doctrine/migrations/pull/162 
- Most of the logs/prints not use psr/log 
- Used dependency injection in most of the "core" code parts (v1 and v2 have the god classes as `Configuration`, `MigrationRepository` and `Version` that were doing all possible kind of things and having circular dependencies between each other )

Tests have not been updated, but the provided prototype has most of the features already working. 
Rollup and Diff commands have not been updated yet.

I realize that is a huge PR and hard to review, but the configuration and migration-repository class was passed all over the places, and as soon as I started to remove methods from it, each other class had to be changed...

Todo:
- [x] migrate all functionalities
- [x] update tests
- [x] phpstan
- [x] prepare migration path for 3.0 update (see https://github.com/goetas/migrations/pull/1)
- [x] document migration path

goetas committed:
use multiple migrations directories and namespaces

goetas committed:
update more commands

goetas committed:
increase version

goetas committed:
allow custom sorting

goetas committed:
executor

goetas committed:
remove start migration

goetas committed:
value object comparison for versions

goetas committed:
adapted execute command

OskarStark commented on lib/Doctrine/Migrations/Metadata/AvailableMigration.php:
```diff
@@ -0,0 +1,33 @@
+use Doctrine\Migrations\AbstractMigration;
+use Doctrine\Migrations\Version\Version;
+
+class AvailableMigration
```
Should be final

goetas replied:
:+1:

OskarStark commented on lib/Doctrine/Migrations/Metadata/AvailableMigration.php:
```diff
@@ -0,0 +1,33 @@
+    /** @var AbstractMigration */
+    private $migration;
+
+    public function __construct(Version $version, AbstractMigration $migration)
```
I am for making the ctor private and a have a static named constructor to be immutable.

Also I am an for version() and migration() without the get prefix

goetas committed:
do not load from sub namespaces

goetas committed:
format migration state

goetas committed:
start adapting tests

goetas committed:
stricter execution result

goetas committed:
create MigratorConfigurationFactory

goetas committed:
extract factory interface

OskarStark commented on lib/Doctrine/Migrations/Version/ExecutionResult.php:
```diff
@@ -49,57 +50,40 @@ class ExecutionResult
-     * @var string
-     */
+    /** @var string */
     private $direction;
```
Is it worth to have a value Direction with Direction::up() and Direction::down() in general?

goetas replied:
I would say yes. It avoids typos and makes clear that there are two possible values without having to look in the abstract migration class.

Here is same debate of enums vs plain strings imo

OskarStark replied:
Glad to hear that

OskarStark commented on lib/Doctrine/Migrations/Metadata/ExecutedMigration.php:
```diff
@@ -0,0 +1,46 @@
+     */
+    public $executionTime;
+
+    public function __construct(Version $version, ?DateTime $executedOn, ?int $executionTime)
```
ExecutionTime::fromMilliseconds(...) ?

goetas replied:
Hmm.. . And that method what should do? Just return back the same value?

OskarStark replied:
Yes but its self explanatory. Looking at `$executionTime` variable name is not helpful, unless it would be named `$executionTimeInMs` oder `$executionTimeInMilliseconds`.

it could also provide a `ExecutionTime->inSeconds()` which could be helpful when outputting it to the user? 🤔

goetas replied:
What you are suggesting is a value object instead of the integer, right?

OskarStark replied:
```php
final class ExecutionTime
{
    /** var int */
    private $time;

    private function __construct(int $time)
    {
        $this->time = $time;
    {

    public function fromMilliseconds(int $time): self
    {
        return new self($time);
    }

    public function toMilliseconds(): int
    {
        return $time;
    }

    public function toSeconds(): // int or float here
    {
        return $time / 1000;
    }
}
```

Something like this

OskarStark commented on lib/Doctrine/Migrations/Metadata/ExecutedMigration.php:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Milliseconds
+     *
+     * @var int
```
Int|null until you add a value object ☺️

goetas replied:
Good idea. I was hoping that somehow doctrine Cs could do it automatically

OskarStark commented on lib/Doctrine/Migrations/Metadata/TableMetadataStorage.php:
```diff
@@ -0,0 +1,100 @@
+                $this->platform->getDateTimeFormatString(),
+                $row['executed_on']
+            );
+            $migration  = new ExecutedMigration($version, $executedOn, $row['execution_time']? intval($row['execution_time']) : null);
```
And here:
```php
ExecutionTime::fromMilliseconds()
```suggestion
            $migration  = new ExecutedMigration($version, $executedOn, $row['execution_time']? ExecutionTime::fromMilliseconds(intval($row['execution_time'])) : null);
```
```

jwage commented:
This is great. Love to see someone tearing this a part more! Can we document the BC breaks and update some of the docs to reflect the changes? I think it would help me understand what has changed so far from the users perspective and would allow me to provide feedback. Let me know when you are ready for a thorough review.

goetas commented:
> This is great. Love to see someone tearing this a part more! Can we document the BC breaks and update some of the docs to reflect the changes? I think it would help me understand what has changed so far from the users perspective and would allow me to provide feedback. Let me know when you are ready for a thorough review.

Sounds great! I have discussed this a bit with @alcaeus , currently I'm spending on this PR 5-8 h/week. I was planning in documenting the changes a bit. As soon as I have something will ping you here!

stof commented:
> * The "version" is not anymore a number or date, but the FQCN of the migration class.

Is there a migration plan available for that ? This involves at least 2 things regarding BC:

- it will probably require to change the schema of the migration table (to allow unlimited length on the version field. How would that schema update be handled ?
- what would happen for an existing project using migrations, where the DB contains dates as versions for already applied migrations. Can these be mapped back to new versions to recognize them as already migrated ?

Also, how will the ordering of migrations work ? Currently, when you have multiple unexecuted migrations, they are applied based on their version using date sorting. But once their version is the FQCN, we need to have a different sorting algorithm, and you don't describe what it is AFAICT.

goetas commented:
> it will probably require to change the schema of the migration table (to allow unlimited length on the version field. How would that schema update be handled ?
what would happen for an existing project using migrations, where the DB contains dates as versions for already applied migrations. Can these be mapped back to new versions to recognize them as already migrated ?

For both of this questions have the same solution. [This](https://github.com/doctrine/migrations/blob/1b44e55200e19d2fcae7eacafb8c1bc8c6640990/lib/Doctrine/Migrations/Metadata/TableMetadataStorage.php#L38) part of code will be responsible for doing that migration. Is not yet coded, but is already on my todo-list.

> Also, how will the ordering of migrations work ? Currently, when you have multiple unexecuted migrations, they are applied based on their version using date sorting. But once their version is the FQCN, we need to have a different sorting algorithm, and you don't describe what it is AFAICT.

The current sorting algorithm used in v2 for migrations is `ksort($migrations,SORT_STRING)` and that will be the same for FQCN, and should produce the same results (in v2 the version is just the part of the class name after Version*). So if migrations are in the same namespace there is no need for any change in the sorting algorithm.
If migrations use multiple namespaces (available only in this proposed 3.0), then you are responsible to provide your sorting algorithm to manage dependencies between namespaces. 
In an ideal world, composer dependency graph could be used to provide out-of-the-box advanced sorting, but this is too app-specific and out of the scope for now.

stof commented:
> n an ideal world, composer dependency graph could be used to provide out-of-the-box advanced sorting, but this is too app-specific and out of the scope for now.

I don't even see how that's possible. your own migrations will still need interleaving with the dependency migrations, as some of your migrations will have been written before the upgrade of one of the dependencies bring a new migration there. A dependency graph is not the solution here.

goetas commented:
If you have single migrations that depend on other migrations from other
packages, you have to manage it on your own with a custom sorting algorithm.

I doubt that will be possible to find a decent way to sort migrations out
of the box and cover most of the cases, thus custom sorting should allow
users to handle it on their own.

This pr does not propose a way to manage migration dependencies, it just
allows you to sort them in any other you prefer.

Did i miss maybe your point or it explains the problem?

On Mon, 16 Sep 2019, 17:52 Christophe Coevoet, <notifications@github.com>
wrote:

> n an ideal world, composer dependency graph could be used to provide
> out-of-the-box advanced sorting, but this is too app-specific and out of
> the scope for now.
>
> I don't even see how that's possible. your own migrations will still need
> interleaving with the dependency migrations, as some of your migrations
> will have been written before the upgrade of one of the dependencies bring
> a new migration there. A dependency graph is not the solution here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/doctrine/migrations/pull/858?email_source=notifications&email_token=AAF5UJ4F7CFJXUNEZO5W5NTQJ6TULA5CNFSM4IWZ62MKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6ZTSIY#issuecomment-531839267>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAF5UJ2X2ZSUDYDGVSCUSVDQJ6TULANCNFSM4IWZ62MA>
> .
>

goetas committed:
fix event dispatcher

goetas committed:
use configuration loaders, drop inheritance

goetas committed:
add config loader tests

goetas committed:
config tests

goetas committed:
executor tests

OskarStark commented on lib/Doctrine/Migrations/DependencyFactory.php:
```diff
@@ -36,17 +41,39 @@ class DependencyFactory
+        return $this->configuration;
+    }
+
+    public function getConnection() : Connection
```
Still needed? For extension?
In the Code I only see $this->connection, same for the hetzet above

goetas committed:
command tests

goetas committed:
up to date command

goetas committed:
migrations repo tests

goetas committed:
metadata tests

goetas committed:
alias resolver tests

goetas committed:
factory test

goetas committed:
cs and tests

goetas committed:
interface

goetas committed:
autocommit listener

goetas commented:
Getting closer to something that works :)
![Code Coverage for  home goetas projects doctrine-migrations lib Doctrine Migrations](https://user-images.githubusercontent.com/776743/65167391-ba4fcf00-da42-11e9-961d-ec4fabf2e902.png)

OskarStark commented on lib/Doctrine/Migrations/Metadata/AvailableMigrationsList.php:
```diff
@@ -51,13 +63,13 @@ public function getMigration(Version $version) : ?AvailableMigration
         }
 
-        return null;
+        throw MigrationNotAvailable::new((string) $version);
```
Maybe MigrationNotAvailable::forVersion() or byVersion() ?

OskarStark replied:
same here

goetas committed:
migration plan calculator tests

goetas committed:
simplify code for migration plan calculator

goetas committed:
more tests

goetas committed:
rollup test

goetas committed:
restore file builder

goetas committed:
test schema diff

goetas committed:
stub schema provider

goetas committed:
dump schema test

goetas committed:
generate test

goetas committed:
diff command

goetas committed:
dix deps

goetas committed:
execute command

goetas committed:
migrate command

goetas committed:
remove outdated tests

goetas commented:
Current status: ~90% of functionalities have been migrated (`--range*` feature on the `migrate` command is still missing). 

I have also migrated most of the tests, wrote some new and deleted a lot of other. This is some critical issue. Since the previous version had 3-4 god classes tightly couples doing most of the job (`Version`, `Configuration`, `MigrationRepository`), there were a bunch of tests checking that the interaction between those are correct. 
In the new implementation, the size of those classes has been cut by ten in some cases, and logic has been split in more independent classes. Because of that, the previously exiting tests became much more of a problem rather than a benefit.. thus I have deleted them.

My plan in the next days is to restore the test coverage to something around 95%, before that I plan to update the documentation with the user-facing changes. Documenting the code changes is a bit more difficult... what about using https://github.com/Roave/BackwardCompatibilityCheck ?

/cc @jwage @alcaeus 

@OskarStark Thanks for the revirews, as soon as I will have a more polished architecture, will start addressing the comments.

jwage commented:
@goetas Sounds like a good plan to me. We also need to at some point look in to how this affects all the 3rd party integrations, like the DoctrineMigrationsBundle for Symfony for example.

goetas commented:
> We also need to at some point look in to how this affects all the 3rd party integrations, like the DoctrineMigrationsBundle for Symfony for example.

That is on my todo-list (since the project that i'm working on will need the bundle too)

goetas committed:
simplify class

goetas committed:
Update documentation

goetas commented:
@jwage https://github.com/doctrine/migrations/pull/858/commits/d916a07c463345d1bcf313342e8b78e320563f96 documents the user-facing changes.  

Some of the documented things are not yet fully implemented (home to have some time next week to discuss them with @alcaeus )

goetas committed:
more changes

goetas committed:
add event arg test

goetas committed:
static analysis

goetas committed:
moved new-migrations and unavailable migrations to the plan calculator

goetas commented:
I've analyzed the current master and this branch with [PhpDependencyAnalysis](https://github.com/mamuz/PhpDependencyAnalysis) to track dependencies between classes in this doctrine/migrations.

The configuration file I've used is:

```yaml
mode: 'usage'
source: './lib'
filePattern: '*.php'
formatter: 'PhpDA\Writer\Strategy\Svg'
target: './complex-cycle.svg'
groupLength: 3
visitorOptions:
  PhpDA\Parser\Visitor\Required\DeclaredNamespaceCollector: {minDepth: 2, sliceLength: 6, excludePattern: '/^((?!.*(Doctrine\\Migrations))|(?=.*Exception)).*$/'}
  PhpDA\Parser\Visitor\Required\MetaNamespaceCollector: {minDepth: 2, sliceLength: 4, excludePattern: '/^((?!.*(Doctrine\\Migrations))|(?=.*Exception)).*$/'}
  PhpDA\Parser\Visitor\Required\UsedNamespaceCollector: {minDepth: 2, sliceLength: 4, excludePattern: '/^((?!.*(Doctrine\\Migrations))|(?=.*Exception)).*$/'}
```

The results are the following https://gist.github.com/goetas/e5ee7546496e673360df5657ec5b38bd .

Red line in the graph are circular dependencies between classes. As you can see, in v3 all circular deps are gone.

goetas committed:
green tests

goetas commented:
`STATIC_ANALYSIS` (phpstan) still missing to be green... could be pretty challenging make it work!

Any help?

stof commented on lib/Doctrine/Migrations/Configuration/Connection/Loader/ConnectionHelperLoader.php:
```diff
@@ -14,30 +14,29 @@
+
+    public function __construct(?HelperSet $helperSet = null, string $helperName, ConnectionLoaderInterface $fallback)
     {
+        $this->helperSet  = $helperSet ?:  new HelperSet();
```
duplicate space after the ternary (weird that this was not catched by CS)

goetas replied:
True, I expected phpcs to auto fix it

stof commented on lib/Doctrine/Migrations/Configuration/Connection/Loader/NoConnectionLoader.php:
```diff
@@ -0,0 +1,23 @@
+use Doctrine\Migrations\Tools\Console\Exception\ConnectionNotSpecified;
+
+/**
+ * The ConnectionConfigurationChainLoader class is responsible for loading a Doctrine\DBAL\Connection from an array of
```
wrong phpdoc. this class is not the ConnectionConfigurationChainLoader

stof commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        $xsdPath = __DIR__ . DIRECTORY_SEPARATOR . 'XML' . DIRECTORY_SEPARATOR . 'configuration.xsd';
+// @todo restore validation
```
what about this TODO ?

goetas replied:
The xsd schema has not been updated yet. Will be done in the next days

stof commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,105 @@
+        if (! file_exists($file)) {
+            throw FileNotFound::new();
+        }
+        libxml_use_internal_errors(true);
```
the existing state must be restored (using a `try/finally`) to avoid changing the global state.

stof replied:
btw, I think 2.x suffers from this too.

stof commented on lib/Doctrine/Migrations/DependencyFactory.php:
```diff
@@ -63,141 +101,169 @@ public function getSchemaDumper() : SchemaDumper
+            return new GlobFinder();
+        });
+
+        // todo move this to DI
```
can it be done ?

goetas replied:
The ideas is that based on the migrations folder schema, here should be passed the right finder

stof commented on lib/Doctrine/Migrations/Generator/FileBuilder.php:
```diff
@@ -65,6 +46,9 @@ public function buildMigrationFile(
 
     private function getVersionUpdateQuery(string $version, string $direction) : string
     {
+        // @todo this is hard to implement since MetadataStorage abstracts the migrations table
```
but without this, things are broken as the metadata storage is not updated with executing this file. So this must be solved.
It might mean that the MetadataStorage needs to expose a method returning the needed queries.

goetas replied:
Does it make sense then to have metadata storage interface? If it is going to be only one implementation, what about dropping the interface and typehint the concrete implementation?

stof replied:
well, the metadata storage interface could have a method returning the queries necessary to update the storage (which may be no queries if the storage cannot be updated in SQL).

goetas replied:
But then metadata storage interface will be bound to SQL, having other "storages"  will make no sense... right? or in that case the extra method for the queries will just return NULL? )

goetas replied:
I have discussed this with @alcaeus . This will be covered in a separate PR/ticket and there can be explored a solution for the issue

stof reviewed (changes requested):
you need to fix the `autoload_directories` in phpstan.neon.dist, as the `_regression` folder is gone. That's what's breaking phpstan entirely right now.

goetas committed:
fix most of the phpstan stuff

stof commented on lib/Doctrine/Migrations/Metadata/AvailableMigrationsList.php:
```diff
@@ -31,16 +32,28 @@ public function getItems() : array
-        return $this->items[count($this->items)-1-(-1*$offset)] ?? null;
+        $offset = count($this->items)-1-(-1*$offset);
+        if (! isset($this->items[$offset])) {
+            throw NoMigrationsFoundWithCriteria::new('first' . ($offset>0 ? ('+' . $offset) : ''));
```
shouldn't we report it as `last -3` rather than `first +41`, as the criteria was based on the last migration ?

stof commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -83,10 +83,14 @@ public function getExecutedMigrations() : ExecutedMigrationsSet
+            $executedAt = (isset($row[$this->configuration->getExecutedAtColumnName()])
+                && $row[$this->configuration->getExecutedAtColumnName()] !== null
+                && $row[$this->configuration->getExecutedAtColumnName()] !== ''
+                ? DateTime::createFromFormat(
```
why using a DateTime and not a DateTimeImmutable ?

stof commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -123,7 +127,7 @@ public function complete(ExecutionResult $result) : void
             $this->connection->insert($this->configuration->getTableName(), [
                 $this->configuration->getVersionColumnName() => (string) $result->getVersion(),
-                $this->configuration->getExecutedAtColumnName() => $result->getExecutedAt() ? $result->getExecutedAt()->format($this->platform->getDateTimeFormatString()): null,
+                $this->configuration->getExecutedAtColumnName() => $result->getExecutedAt()!== null ? $result->getExecutedAt()->format($this->platform->getDateTimeFormatString()): null,
```
this could be simplified by using the `$types` third argument instead, and letting DBAL perform the conversion

stof commented on lib/Doctrine/Migrations/MigrationPlanCalculator.php:
```diff
@@ -52,12 +53,12 @@ public function getPlanUntilVersion(?Version $to = null) : MigrationPlanList
+        try {
+            $to = $to ?: $availableMigrations->getLast()->getVersion();
+        } catch (NoMigrationsFoundWithCriteria $e) {
             throw NoMigrationsToExecute::new();
```
should we attach the catched exception as previous exception ?

stof commented on lib/Doctrine/Migrations/Tools/Console/Command/AbstractCommand.php:
```diff
@@ -75,26 +77,39 @@ protected function outputHeader(
-
-            $logger                  = new ConsoleLogger($output);
-            $this->dependencyFactory = new DependencyFactory($configuration, $connection, $em, $logger);
+            // todo doctrine exception
```
should be done.

goetas committed:
update xsd validator for 3.0

goetas committed:
phpstan

goetas commented:
Only 111 errors left :-)

goetas committed:
add constructor provided migrations

goetas committed:
add missing range feature

goetas committed:
fix some failing tests

goetas committed:
missing factory method

goetas committed:
add config loader

goetas committed:
first/last tests

goetas committed:
di test

leonhusmann commented on lib/Doctrine/Migrations/Configuration/Configuration.php:
```diff
@@ -174,23 +190,45 @@ public function getQuotedMigrationsExecutedAtColumnName() : string
     }
 
+    /**
+     * @deprecated Use getMigrationsDirectory()
```
Looks wrong to me, when this Method is deprecated I cannot use the same method instead.

goetas replied:
Hi, Thanks for the review, but your comment is based on an outdated file (as highlighted by github).
I suggest to do a review on the latest version of this PR visible from the [full diff](https://github.com/doctrine/migrations/pull/858/files) or to checkout this PR locally by cloning https://github.com/goetas/migrations.

goetas committed:
move migrator option to command

goetas committed:
migrator tests

goetas committed:
phpstan and tests

goetas committed:
restore down migration behaviour

goetas commented:
PHPStan 0 errors! :tada: :tada: :tada:

goetas committed:
stof code review suggestions

goetas committed:
use datetime immutable

goetas committed:
mark value objects as final

goetas committed:
single queries are run in debug mode

goetas commented:
Removing the WIP state from the pull request.

What is left to do:
- fix windows tests
- prepare migration path for 3.0 update 
   -- automate migration of the v1/v2 migrations table (table metadata storage can do that in a relatively easy way)
- discuss what to do with the `down()` method
- create compatibility/deprecation layer in v2
- test with real projects and get more feedback!!

goetas committed:
ensure proper typecasting

goetas committed:
skip migration version table in the schema dumps

goetas committed:
allow to filter tables in a dump

goetas committed:
prepare upgrading document

goetas commented:
Pushed also an upgrading document

goetas commented:
https://github.com/goetas/migrations/pull/1 is a proposal on how to auto-migrate the version table from V1 or V2 to V3.

goetas committed:
version it is a string

goetas committed:
fix too large column size

goetas committed:
change output levels

alcaeus commented on lib/Doctrine/Migrations/Tools/Console/Command/DiffCommand.php:
```diff
@@ -82,7 +71,7 @@ protected function configure() : void
                 InputOption::VALUE_OPTIONAL,
                 'Check Database Platform to the generated code.',
-                true
+                1
```
Why `int` instead of `bool`? This command also shouldn't have a value as it's just a switch IIRC.

goetas replied:
phpstan complains

`Parameter #5 $default of method Symfony\Component\Console\Command\Command::addOption() expects int|string|false|null, true given.`

alcaeus replied:
Ok. That's something we should investigate separately.

alcaeus commented on lib/Doctrine/Migrations/Tools/Console/ConsoleLogger.php:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @see https://www.php-fig.org/psr/psr-3/
+ */
+class ConsoleLogger extends AbstractLogger
```
```suggestion
final class ConsoleLogger extends AbstractLogger
```

Shouldn't this also be an internal class?

goetas replied:
true

alcaeus commented on lib/Doctrine/Migrations/Tools/Console/Helper/ConfigurationHelper.php:
```diff
@@ -22,37 +21,24 @@
  */
 class ConfigurationHelper extends Helper implements ConfigurationHelperInterface
```
```suggestion
final class ConfigurationHelper extends Helper implements ConfigurationHelperInterface
```

alcaeus commented on lib/Doctrine/Migrations/Tools/Console/MigratorConfigurationFactory.php:
```diff
@@ -0,0 +1,32 @@
+
+    public function getMigratorConfiguration(InputInterface $input) : MigratorConfiguration
+    {
+        $timeAllQueries = $input->hasOption('query-time') ?(bool) $input->getOption('query-time'): false;
```
```suggestion
        $timeAllQueries = $input->hasOption('query-time') ? (bool) $input->getOption('query-time') : false;
```

alcaeus commented on lib/Doctrine/Migrations/Tools/Console/MigratorConfigurationFactoryInterface.php:
```diff
@@ -0,0 +1,13 @@
+use Doctrine\Migrations\MigratorConfiguration;
+use Symfony\Component\Console\Input\InputInterface;
+
+interface MigratorConfigurationFactoryInterface
```
```suggestion
interface MigratorConfigurationFactory
```

alcaeus commented on lib/Doctrine/Migrations/Tools/Console/MigratorConfigurationFactory.php:
```diff
@@ -0,0 +1,32 @@
+use Doctrine\Migrations\MigratorConfiguration;
+use Symfony\Component\Console\Input\InputInterface;
+
+class MigratorConfigurationFactory implements MigratorConfigurationFactoryInterface
```
```suggestion
class ConsoleInputMigratorConfigurationFactory implements MigratorConfigurationFactory
```

alcaeus commented on lib/Doctrine/Migrations/Version/Factory.php:
```diff
@@ -14,25 +16,23 @@
  */
 class Factory
```
```suggestion
final class Factory
```

alcaeus commented on lib/Doctrine/Migrations/Version/Factory.php:
```diff
@@ -14,25 +16,23 @@
     }
 
-    public function createVersion(string $version, string $migrationClassName) : Version
+    public function createVersion(string $migrationClassName) : AbstractMigration
```
I'm confused: this class is in the `Version` namespace but returns a migration. Should we rename it `MigrationFactory` instead of just `Factory`?

alcaeus commented on phpcs.xml.dist:
```diff
@@ -35,18 +39,36 @@
+        <exclude-pattern>tests/Doctrine/Migrations/Tests/TestLogger.php</exclude-pattern>
+    </rule>
+
+    <rule ref="SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingReturnTypeHint">
```
What's the reason to exclude all these files?

goetas replied:
Fixed. Only `lib/Doctrine/Migrations/Version/Version.php` is needed because of __toStirng method

alcaeus commented on phpcs.xml.dist:
```diff
@@ -58,6 +80,11 @@
         <exclude-pattern>lib/Doctrine/Migrations/ParameterFormatterInterface.php</exclude-pattern>
-        <exclude-pattern>lib/Doctrine/Migrations/Version/VersionInterface.php</exclude-pattern>
+        <exclude-pattern>lib/Doctrine/Migrations/MigratorInterface.php</exclude-pattern>
+        <exclude-pattern>lib/Doctrine/Migrations/Tools/Console/MigratorConfigurationFactoryInterface.php</exclude-pattern>
```
This becomes obsolete if you rename the interface above.

alcaeus reviewed (commented):
I've reviewed code changes and tried to keep style-related comments to a minimum.

There are some very good changes here, especially in terms of cleaning up some legacy behaviour and legacy quirks.

I'll let you decide which comments you want to resolve in this PR and which you'd rather handle in a separate PR. Just let me know in the respective comments.

I've added comments to make some classes final. I've refrained from adding comments to rename classes as I'd rather not deal with that number of BC breaks in this release. We can always rename and re-organise code in future minor versions with a good compatibility layer (e.g. moving all the Console stuff out of the `Tools` namespace).

Thanks for leading this huge effort! ❤️

goetas commented on UPGRADE.md:
```diff
@@ -1,3 +1,280 @@
+# Upgrade to 3.0
```
I personally prefer to keep it. But my opinion is not that strong on it, so if you prefer I can just drop it.

I can understand symfony that in some cases maintains 3 or more branches in parallel... thus the "upgrade" might be updated in the various branches.

alcaeus replied:
We'll leave it for now and do that in a separate PR.

goetas commented on docs/en/reference/configuration.rst:
```diff
@@ -95,17 +126,17 @@ Here are details about what each configuration option does:
+| table_storage.version_column_name                | no         | version                      | The name of the column which stores the version name.                            |
 +----------------------------+------------+------------------------------+----------------------------------------------------------------------------------+
-| column_length              | no         | 14                           | The length of the column which stores the version name.                          |
+| table_storage.version_column_length              | no         | 2048                           | The length of the column which stores the version name.                          |
```
in reality is 1024, will fix it. it is due to the usage of FQCN.

goetas commented on lib/Doctrine/Migrations/Configuration/Loader/AbstractFileLoader.php:
```diff
@@ -0,0 +1,32 @@
+use function realpath;
+
+/**
+ * The ArrayConfiguration class is responsible for loading migration configuration information from a PHP file.
```
dropped the description

goetas commented on lib/Doctrine/Migrations/Configuration/Loader/ArrayLoader.php:
```diff
@@ -0,0 +1,106 @@
+ *
+ * @internal
+ */
+class ArrayLoader implements Loader
```
makes sense,

OskarStark commented:
@alcaeus I like your strictness via final classes 👍🏻
I didn’t check, but methods which are now protected should be private afterwards if the class is final and the method is not overwritten from a base/abstract class which enforces this visibility

goetas commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,116 @@
+        $rawXML = file_get_contents($file);
+        assert($rawXML !== false);
+
+        $root = simplexml_load_string($rawXML, SimpleXMLElement::class, LIBXML_NOCDATA);
```
simplexml_load_string does not suppor all the "stream wrappers" that file_get_contents does

alcaeus replied:
I see. Thanks for the explanation

OskarStark commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -0,0 +1,140 @@
+    {
+        $this->connection->executeUpdate(
+            sprintf(
+                'DELETE FROM %s WHERE 1 = 1',
```
@Ocramius do you have any experience on this?

goetas replied:
recently sql databases started to now allow delete and update with no conditions... very annoyoig.

alcaeus replied:
This was more a question of curiosity - no problem keeping this for now if it causes trouble otherwise.

OskarStark commented on lib/Doctrine/Migrations/Metadata/ExecutedMigrationsSet.php:
```diff
@@ -0,0 +1,79 @@
+    {
+        $offset = count($this->items)-1-(-1*$offset);
+        if (! isset($this->items[$offset])) {
+            throw NoMigrationsFoundWithCriteria::new('last' . ($offset>0 ? ('+' . $offset) : ''));
```
Couldn’t we use a named constructor for this and do the string concat stuff inside?

OskarStark commented on lib/Doctrine/Migrations/Finder/RecursiveRegexFinder.php:
```diff
@@ -14,8 +14,20 @@
+    public function __construct(?string $pattern = null)
+    {
+        $this->pattern = $pattern ?: sprintf(
+            '#^.+\\%sVersion[^\\%s]{1,255}\\.php$#i',
```
👍🏻 for the warning

goetas replied:
What about allowing any class name? (since the sorting is not a problem anymore?)
Consider that this is not a limitation I have introduced, it was already present in 2.x (see https://github.com/doctrine/migrations/blob/1f9e877ef5a9b5403fc6581449f3fe55c8518500/lib/Doctrine/Migrations/Finder/RecursiveRegexFinder.php#L47)

alcaeus replied:
Yeah, that's what I thought. If there's no reason to keep the limit, let's drop it.

goetas replied:
done

goetas commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -0,0 +1,140 @@
+        $this->schemaManager = $connection->getSchemaManager();
+        $this->platform      = $connection->getDatabasePlatform();
+
+        if ($configuration!== null && ! ($configuration instanceof TableMetadataStorageConfiguration)) {
```
I'm also very surprised

goetas commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -0,0 +1,140 @@
+    {
+        if ($this->connection instanceof MasterSlaveConnection) {
+            $this->connection->connect('master');
+        }
```
This was already there, but the issue is that we need to make sure that the "select" that gets the list of executed migrations come from the "master". 

By default DBAL chooses the "master" only on the first "update", in this way we force it even on selects

alcaeus commented:
@OskarStark "shoot first, ask questions later": we can always remove `final` in a future minor release while the opposite is nigh impossible. That's why I've gotten used to only having abstract or final classes if it can be helped.

Regarding protected methods, you're absolutely right. I've taken your suggestion and suggested it for inclusion in slevomat/coding-standard, which serves as the base for doctrine/coding-standard: https://github.com/slevomat/coding-standard/issues/754. Maybe I can find the time to work on it soon, as I think this makes sense to check automatically.

OskarStark commented:
Definitely +final 👌🏻👍🏻

At CS: 👍🏻
I am using g Php-cs-fixer which can do this already

goetas commented on lib/Doctrine/Migrations/Configuration/Exception/UnknownLoader.php:
```diff
@@ -0,0 +1,16 @@
+{
+    public static function new(string $loader) : self
+    {
+        return new self(sprintf('Unknown configuration loader "%s".', $loader), 10);
```
Exception codes were used in previous exception classes... will be happy to drop them

goetas commented on lib/Doctrine/Migrations/Generator/SqlGenerator.php:
```diff
@@ -26,13 +28,16 @@ class SqlGenerator
     private $platform;
 
-    public function __construct(Configuration $configuration, AbstractPlatform $platform)
+    public function __construct(Configuration $configuration, AbstractPlatform $platform, MetadataStorageConfigration $storageConfiguration)
```
my mistake. fixed

goetas committed:
fixed a lot of stuff from code review

goetas committed:
move generateVersionNumber out of config class

goetas committed:
rename MigratorConfigurationFactory

goetas committed:
rename Factory

goetas committed:
cs

goetas committed:
add version equals method

goetas commented on lib/Doctrine/Migrations/Metadata/AvailableMigrationsList.php:
```diff
@@ -0,0 +1,80 @@
+    public function hasMigration(Version $version) : bool
+    {
+        foreach ($this->items as $migration) {
+            if ((string) $migration->getVersion() === (string) $version) {
```
We were talking that "comparable" does not work and I was saying that "equals" should be implemented by php itself. Anyway, I have added it now.

OskarStark commented on lib/Doctrine/Migrations/Configuration/Loader/PHPFileLoader.php:
```diff
@@ -0,0 +1,49 @@
+    public function load($file) : Configuration
+    {
+        if (! file_exists($file)) {
+            throw FileNotFound::new();
```
What about something like this?
```suggestion
            throw FileException::notFound();
```

all the time `new()` feels a bit weird.

goetas replied:
are you suggesting to drop the `FileNotFound` class in favor of a more generic `FileException`?

OskarStark commented on lib/Doctrine/Migrations/Metadata/AvailableMigrationsList.php:
```diff
@@ -0,0 +1,80 @@
+    public function getFirst(int $offset = 0) : AvailableMigration
+    {
+        if (! isset($this->items[$offset])) {
+            throw NoMigrationsFoundWithCriteria::new('first' . ($offset > 0 ? ('+' . $offset) : ''));
```
or sth like:
```suggestion
            throw NoMigrationsFound::withCriteria('first' . ($offset > 0 ? ('+' . $offset) : ''));
```

localheinz commented on lib/Doctrine/Migrations/AbstractMigration.php:
```diff
@@ -73,11 +70,7 @@ public function warnIf(bool $condition, string $message = '') : void
-            $this->version->getExecutionState(),
-            $message
-        ));
+        $this->logger->warning($message, ['migration' => $this]);
```
Since `$context` allows for more data, in order to keep following diffs smaller, what do you think about wrapping the `$context` argument?

```suggestion
        $this->logger->warning($message, [
            'migration' => $this,
        ]);
```

goetas replied:
this is currently managed by doctrine-coding standards, and wants it in-line

alcaeus replied:
That is correct - doctrine/cs forces single-line for arrays with a single element.

localheinz commented on lib/Doctrine/Migrations/Configuration/ConfigurationLoader.php:
```diff
@@ -0,0 +1,50 @@
+    private function setDefaultLoaders() : void
+    {
+        $this->loaders = [
+            'array' => new ArrayLoader(),
```
Is the order relevant here?

If not, what do you think about keeping this `array` sorted by key?

localheinz commented on lib/Doctrine/Migrations/Configuration/Connection/ConnectionLoaderInterface.php:
```diff
@@ -17,6 +18,8 @@ interface ConnectionLoaderInterface
+     * @throws ConnectionNotSpecified
      */
-    public function chosen() : ?Connection;
+    public function getConnection() : Connection;
```
Do we need the `get` prefix here or can we drop it?

goetas replied:
For consistency with the res of doctrine libraries will keep it as `getConnection` ... see ORM, DBAL, and this library too..

localheinz commented on lib/Doctrine/Migrations/Configuration/ConfigurationLoader.php:
```diff
@@ -0,0 +1,50 @@
+use Doctrine\Migrations\Configuration\Loader\YamlFileLoader;
+use function count;
+
+class ConfigurationLoader
```
Can this class be `final`?

goetas replied:
marking it as final will make `ConfigurationHelperTest` quite more complex. But thanks to your review i noticed that i forgot to mark it as internal. Thanks

localheinz commented on lib/Doctrine/Migrations/Configuration/Connection/Loader/ConnectionHelperLoader.php:
```diff
@@ -14,30 +14,29 @@
+    /** @var ConnectionLoaderInterface */
+    private $fallback;
+
+    public function __construct(?HelperSet $helperSet = null, string $helperName, ConnectionLoaderInterface $fallback)
```
Can we move parameters with a default value to the end of the parameter list, or better even, drop the default value?

Also see 

* https://github.com/localheinz/phpstan-rules#methodsnoparameterwithnullabletypedeclarationrule
* https://github.com/localheinz/phpstan-rules#methodsnoparameterwithnulldefaultvaluerule

localheinz commented on lib/Doctrine/Migrations/Configuration/Exception/UnknownResource.php:
```diff
@@ -0,0 +1,16 @@
+{
+    public static function new(string $loader) : self
+    {
+        return new self(sprintf('The provided resource can not be loaded by the loader "%s".', $loader), 10);
```
Perhaps 

* the exception class should be named `UnableToLoadResource` and 
* the named constructor should be named `with(string $loader)`?

?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/AbstractFileLoader.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return array<string,string>
+     */
+    protected function getDirectoryRelativeToFile(string $file, array $input) : array
```
Should this method be `final`?

Also see https://github.com/localheinz/phpstan-rules#methodsfinalinabstractclassrule.

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/ArrayLoader.php:
```diff
@@ -0,0 +1,106 @@
+            }
+
+            if (is_array($configMap[$configurationKey])) {
+                if ($configurationKey === 'table_storage') {
```
I'd suggest to be as strict as possible - ignoring an unknown configuration option might suggest to the configuring developer that the option has been accepted and had an effect.

goetas replied:
done

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/JsonFileLoader.php:
```diff
@@ -0,0 +1,58 @@
+    /** @var ArrayLoader */
+    private $arrayLoader;
+
+    public function __construct(?ArrayLoader $arrayLoader = null)
```
It's obviously a hard dependency, what do you think about making that clear by removing both the

- nullable type declaration and the
- default value

?

goetas replied:
set as hard dependency

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/JsonFileLoader.php:
```diff
@@ -0,0 +1,58 @@
+    public function load($file) : Configuration
+    {
+        if (! file_exists($file)) {
+            throw FileNotFound::new();
```
Perhaps the named constructor should accept the name of the file here?

```suggestion
            throw FileNotFound::file($file);
```

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/JsonFileLoader.php:
```diff
@@ -0,0 +1,58 @@
+        $config = json_decode($contents, true);
+
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw JsonNotValid::new();
```
Perhaps we should have a dedicated named constructor?

```suggestion
            throw JsonNotValid::file($file);
```

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/PHPFileLoader.php:
```diff
@@ -0,0 +1,49 @@
+/**
+ * @internal
+ */
+final class PHPFileLoader extends AbstractFileLoader
```
Should this be `PhpFileLoader` instead?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/PHPFileLoader.php:
```diff
@@ -0,0 +1,49 @@
+    /** @var ArrayLoader */
+    private $arrayLoader;
+
+    public function __construct(?ArrayLoader $arrayLoader = null)
```
It's obviously a hard dependency, what do you think about making that clear by removing both the

- nullable type declaration and the
- default value

?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/PHPFileLoader.php:
```diff
@@ -0,0 +1,49 @@
+    public function load($file) : Configuration
+    {
+        if (! file_exists($file)) {
+            throw FileNotFound::new();
```
Or even

```suggestion
            throw InvalidFile::notFound($file);
```

?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,126 @@
+    /** @var ArrayLoader */
+    private $arrayLoader;
+
+    public function __construct(?ArrayLoader $arrayLoader = null)
```
It's obviously a hard dependency, what do you think about making that clear by removing both the

- nullable type declaration and the
- default value

?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,126 @@
+    public function load($file) : Configuration
+    {
+        if (! file_exists($file)) {
+            throw FileNotFound::new();
```
See earlier comment.

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * @return mixed[]
+     */
+    private function extractParameters(SimpleXMLElement $root, bool $loopOverNodes = true) : array
```
Can we drop the default value and instead explicitly specify at call sites?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,126 @@
+        return $config;
+    }
+
+    private function validateXML(string $file) : void
```
For consistency, should this be named

```suggestion
    private function validateXml(string $file) : void
```

?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php:
```diff
@@ -0,0 +1,126 @@
+            $xsdPath = __DIR__ . DIRECTORY_SEPARATOR . 'XML' . DIRECTORY_SEPARATOR . 'configuration.xsd';
+
+            if ($xml->schemaValidate($xsdPath) === false) {
+                throw XmlNotValid::failedValidation();
```
Perhaps

```suggestion
                throw XmlNotValid::file($file, $xsdPath);
```

?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/YamlFileLoader.php:
```diff
@@ -2,33 +2,46 @@
+    /** @var ArrayLoader */
+    private $arrayLoader;
+
+    public function __construct(?ArrayLoader $arrayLoader = null)
```
It's obviously a hard dependency, what do you think about making that clear by removing both the

- nullable type declaration and the
- default value

?

localheinz commented on lib/Doctrine/Migrations/Configuration/Loader/YamlFileLoader.php:
```diff
@@ -2,33 +2,46 @@
         }
 
+        if (! file_exists($file)) {
+            throw FileNotFound::new();
```
See comment above.

localheinz reviewed (commented):
Quite a lot to review, cannot manage in a single session.

goetas commented on lib/Doctrine/Migrations/AbstractMigration.php:
```diff
@@ -32,18 +29,18 @@ abstract class AbstractMigration
-    {
-        $config = $version->getConfiguration();
+    /** @var mixed[] */
+    private $plannedSql = [];
```
In this context "planned" is important, because that is the list of sql commands that this migration wants to run.

localheinz replied:
The accessor is called `getSql()` - should it also have a `Planned` infix, then?

Or perhaps it is clear already that a migration will only provide SQL statements that need to be run?

goetas committed:
rename interfaces

goetas committed:
more readable executed at code

goetas committed:
ensure lowercase column names

goetas commented on lib/Doctrine/Migrations/Metadata/ExecutedMigrationsSet.php:
```diff
@@ -0,0 +1,79 @@
+use function array_values;
+use function count;
+
+final class ExecutedMigrationsSet implements Countable
```
Because of type hinting this is not really possible. Each of the duplicated methods has a different signature...

alcaeus replied:
Ah, I understand. Let's change this once PHP supports generics 😂

goetas committed:
document migrator interface

goetas commented on lib/Doctrine/Migrations/SchemaDumper.php:
```diff
@@ -60,6 +75,10 @@ public function dump(
         $down = [];
 
         foreach ($schema->getTables() as $table) {
+            if ($this->shouldSkipTable($table, $excludedTablesRegexes)) {
```
Extra exclusion regex can be provided via console `--filter-tables` param (this is a new feature).

In our case will allow us to dump different postgres table schema to different migration namespaces

goetas commented:
@alcaeus I have addressed ~80% of your review suggestions. 
- The one that are just marked as resolved have been implemented as suggested
- The one that in have just commented but not marked as resolved, are waiting for feedback from you
- The one that do not have yet a feedback from me are stil to be done

@OskarStark @localheinz A huge than you for your reviews. Some of them are already implemented (see as above), but many of your change request are about code style and renaming methods/classes or some strict coding style. I agree with most of your suggestion but definitively don't have the time to implement all of them. 
I have a suggestion. What about making a PR to my branch with the proposed changes? (in alternative if this is not ok for you we can wait for this PR to be merged and than do a second round of PRs to address those issues.

greg0ire commented on UPGRADE.md:
```diff
@@ -1,3 +1,282 @@
+# Upgrade to 3.0
+
+- The "version" is the FQCN of the migration class (exiting entries in the migrations table will be automatically updated).
```
```suggestion
- The "version" is the FQCN of the migration class (existing entries in the migrations table will be automatically updated).
```

greg0ire commented on UPGRADE.md:
```diff
@@ -1,3 +1,282 @@
+# Upgrade to 3.0
+
+- The "version" is the FQCN of the migration class (exiting entries in the migrations table will be automatically updated).
+- `MigrationsEventArgs` and `MigrationsVersionEventArgs` expose different API, please refer to the *Code BC breaks* section.
```
Please add a link, Github creates anchors for every title

greg0ire commented on UPGRADE.md:
```diff
@@ -1,3 +1,282 @@
+- `MigrationsEventArgs` and `MigrationsVersionEventArgs` expose different API, please refer to the *Code BC breaks* section.
+
+## Console
+- Console output changed. The commands use different output style. If you were relying on specific output, 
```
```suggestion
- Console output changed. The commands use a different output style. If you were relying on a specific output, 
```

greg0ire commented on UPGRADE.md:
```diff
@@ -1,3 +1,282 @@
+## Console
+- Console output changed. The commands use different output style. If you were relying on specific output, 
+  please update your scripts. 
+  Console output is not covered by the BC promise, so please try not to rely on specific output.
```
```suggestion
  Console output is not covered by the BC promise, so please try not to rely on a specific output.
```

greg0ire commented on UPGRADE.md:
````diff
@@ -1,3 +1,282 @@
+    'check_database_platform' => true,
+];
+```
+In similar way also XML, YAML and JSON file changed. Please refer to the official documentation for more details.
````
```suggestion
Files in XML, YAML or JSON also changed in a similar way. Please refer to the official documentation for more details.
```

greg0ire commented on UPGRADE.md:
```diff
@@ -1,3 +1,282 @@
+
+## Code BC breaks
+
+Most of the code is protected by the `@internal` declaration and in a very rare case you should have dealt with the 
```
```suggestion
Most of the code is protected by the `@internal` declaration and in a very rare cases you might have dealt with the 
```

greg0ire commented on .appveyor.yml:
```diff
@@ -56,7 +56,6 @@ install:
 
     # install composer dependencies
     - cd C:\projects\migrations
-    - rm composer.lock
```
This file was modified in a commit labeled "appveyor". Please rewrite your commit history and include an explanation about this in your commit message :pray:

goetas replied:
https://github.com/doctrine/migrations/commit/12d408e7caa5fb9f5e921152b8ee6cfd4cc72d64#diff-354f30a63fb0907d4ad57269548329e3R19 this added the idea of commiting the composer.lock file in the repo, and from my understanding, it is a very bad idea.

Some reasons:
1) tests are run with outdated deps (as example some CS failures were not discovered because of it)
2) devs when contributing, and running composer install on their computers might get dependencies not compatible with their dev environment


The only reason I see to commit the composer.lock is to run tests faster... was it.

@jwage @alcaeus I can put it back, but should be clear which PHP version are we targeting

greg0ire replied:
I don't like that policy either, but the actual reason is documented here: https://www.doctrine-project.org/policies.html

I think we shouldn't remove the `composer.lock` from version control without amending or removing that policy page beforehand.

goetas replied:
reverted in https://github.com/doctrine/migrations/pull/858/commits/4eda5e6f78b2bc620adb56fea2423cfe47d3c1f1

greg0ire commented on composer.json:
```diff
@@ -24,10 +24,11 @@
+        "phpstan/phpstan-deprecation-rules": "^0.11",
+        "phpstan/phpstan-phpunit": "^0.11",
+        "phpstan/phpstan-strict-rules": "^0.11",
+        "phpstan/phpstan-symfony": "^0.11.6",
```
If you think 473d1bb16bf13355546e2936c3f8e7997983ae1b could be extracted in a separate PR that's something I could have people try to do at an Hacktoberfest event next week. Might help with making that PR smaller. Please tell me if there are any other parts you think would be easy to extract in separate PRs.

goetas replied:
PHPStan has been a major thing in this refactoring... i did a lot of iterations in fixing PHPStan complains... not sure how proceed here

greg0ire replied:
Ok so it's probably too big for that Hacktoberfest event then, nevermind.

greg0ire commented on lib/Doctrine/Migrations/SchemaDumper.php:
```diff
@@ -95,9 +113,23 @@ public function dump(
+    private function shouldSkipTable(Table $table, array $excludedTablesRegexes) : bool
+    {
+        foreach (array_merge($excludedTablesRegexes, $this->excludedTablesRegexes) as $regex) {
+            if (preg_match($regex, $table->getName()) !== 0) {
```
Should there be preg error handling here? Or are the regexes already validated at this point?

goetas replied:
No, regex comes out of a console command line parameter. Which validation should be performed? how?

greg0ire replied:
"validated" is probably a poor choice of words on my part. I was thinking about something like this: https://github.com/symfony/symfony/blob/62216ea67762b18982ca3db73c391b0748a49d49/src/Symfony/Component/Yaml/Parser.php#L1072-L1090

goetas replied:
Done in https://github.com/doctrine/migrations/pull/858/commits/dc174a439bc01baa7571bdcdc173e935b4fbf280 . Are you ok with it?

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/Command/DiffCommand.php:
```diff
@@ -106,10 +104,21 @@ public function execute(
+        if ($namespace === null) {
+            $namespace = key($dirs);
+        } elseif (! isset($dirs[$namespace])) {
+            throw new Exception(sprintf('Path not defined for the namespace %s', $namespace));
```
```suggestion
            throw new OutOfBoundsException(sprintf('Path not defined for the "%s" namespace', $namespace));
```

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/Command/DumpSchemaCommand.php:
```diff
@@ -51,6 +55,18 @@ protected function configure() : void
+                'namespace',
+                null,
+                InputOption::VALUE_REQUIRED,
+                'Namespace for the save the generated migrations (defaults to the first namespace definition).'
```
```suggestion
                'Namespace to use for generated migrations (defaults to the first namespace definition).'
```

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/Command/MigrateCommand.php:
```diff
@@ -196,60 +213,32 @@ private function checkExecutedUnavailableMigrations(
+                $output->writeln(sprintf(
+                    '<error>Unknown version: %s</error>',
+                    OutputFormatter::escape($versionAlias)
+                ));
```
How about:
```php
 if ($versionAlias === 'first') {
    $output->writeln('<error>Already at first version.</error>');
    return;
}
if ($versionAlias === 'next' || $versionAlias === 'latest') {
    $output->writeln('<error>Already at latest version.</error>');
    return;
}
if (substr($versionAlias, 0, 7) === 'current') {
    $output->writeln('<error>The delta could not be reached.</error>');
}
return;

$output->writeln(sprintf(
    '<error>Unknown version: %s</error>',
    OutputFormatter::escape($versionAlias)
));
```

goetas replied:
:+1:

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/Command/MigrateCommand.php:
```diff
@@ -106,88 +109,102 @@ protected function configure() : void
+
+        /**
+         * If there are no migrations to execute throw an exception.
+         */
```
It does not look like an exception is thrown

goetas replied:
True, just status 1 is returned

greg0ire commented on lib/Doctrine/Migrations/Version/AliasResolver.php:
```diff
@@ -41,29 +55,66 @@ public function __construct(MigrationRepository $migrationRepository)
+                try {
+                    return $newMigrations->getFirst()->getVersion();
+                } catch (NoMigrationsFoundWithCriteria $e) {
+                    throw NoMigrationsToExecute::new();
```
Should `$e` end up as `$previous` in the new exception, here and on line 97?

greg0ire commented on lib/Doctrine/Migrations/Version/AliasResolverInterface.php:
```diff
@@ -0,0 +1,28 @@
+     * - next: The version following the current version.
+     * - latest: The latest available version.
+     *
+     * If an existing version number is specified, it is returned verbatimly.
```
```suggestion
     * If an existing version number is specified, it is returned verbatim.
```

greg0ire commented on lib/Doctrine/Migrations/Version/Executor.php:
```diff
@@ -183,161 +171,162 @@ private function startMigration(
-            $this->connection->rollBack();
+        if ($result->isSkipped()) {
+            $this->logger->error(
+                'Migration {version} skipped during %s. Reason {error}',
```
```suggestion
                'Migration {version} skipped during %s. Reason: "{error}"',
```

goetas replied:
:+1:

greg0ire commented on lib/Doctrine/Migrations/Version/Executor.php:
```diff
@@ -183,161 +171,162 @@ private function startMigration(
+            );
+        } elseif ($result->hasError()) {
+            $this->logger->error(
+                'Migration {version} failed during %s. Error {error}',
```
```suggestion
                'Migration {version} failed during %s. Error: "{error}"',
```

goetas replied:
:+1:

greg0ire commented on lib/Doctrine/Migrations/Version/Executor.php:
```diff
@@ -183,161 +171,162 @@ private function startMigration(
-            $this->connection->rollBack();
+        if ($result->isSkipped()) {
+            $this->logger->error(
+                'Migration {version} skipped during %s. Reason {error}',
```
Shouldn't the %s be replaced with `{state}` here and below?

goetas replied:
good catch, fixed

goetas committed:
code reviews by @greg0ire

goetas committed:
alphabetical sorting of loaders

goetas committed:
change argument order of ConnectionHelperLoader

goetas committed:
rename PhpFileLoader

goetas committed:
remove param default

goetas committed:
mark getDirectoryRelativeToFile as final

goetas committed:
rename UnableToLoadResource

goetas committed:
mark ConfigurationLoader as internal

goetas committed:
split table storage documentation

goetas committed:
MigrationNotAvailable::forVersion

goetas commented on docs/en/reference/configuration.rst:
```diff
@@ -95,17 +126,17 @@ Here are details about what each configuration option does:
+| migrations_paths<string, string>       | yes        | null                         | The PHP namespace your migration classes are located under and the path to a directory where to look for migration classes.                     |
 +----------------------------+------------+------------------------------+----------------------------------------------------------------------------------+
-| table_name                 | no         | doctrine_migration_versions  | The name of the table to track executed migrations in.                           |
+| table_storage.table_name                 | no         | doctrine_migration_versions  | The name of the table to track executed migrations in.                           |
```
done

goetas commented on lib/Doctrine/Migrations/Metadata/MigrationPlan.php:
```diff
@@ -0,0 +1,54 @@
+        return $this->result;
+    }
+
+    public function setResult(ExecutionResult $result) : void
```
I would like to make it immutable, but how? by setting some boolean flag and not allow write operations or else?

alcaeus replied:
Yeah. `markAsExecuted` would throw an exception if an execution result has already been set.

goetas commented on lib/Doctrine/Migrations/MigrationRepository.php:
```diff
@@ -37,379 +24,139 @@
+    public function registerMigrationInstance(Version $version, AbstractMigration $migration) : AvailableMigration
     {
-        $this->versions[$version->getVersion()] = $version;
+        // todo mark private
```
I have it on my list, currently is public mainly for easing the testing... marking it private will have to rewrite really a lot of tests... I would suggest to mark it as internal and put a big warning to not use it.... 

other ideas?

alcaeus replied:
No, works for me. I just wanted to highlight the todo comment

goetas commented:
Thanks again everyone for the reviews.I did one more round of fixes. 

I have adopted the same strategy as explained in https://github.com/doctrine/migrations/pull/858#issuecomment-541304319 to mark the review as done,resolved or waiting for feedback. 
For the reviews where I have left an answer and you are satisfied with it, could you please mark it as "resolved"?

goetas committed:
update composer lock

goetas committed:
handle possible regex errors

goetas committed:
diff command tests fixed after rebase

goetas commented:
I have rebased this PR on the latest master. The current branch is able to merge with fast-forward, but github complains that there are conflicts... any idea here?

goetas committed:
upgrade

goetas committed:
run cs on php 7.2

goetas committed:
windows tests

goetas commented:
I think that windows tests are not working even on the current master...

alcaeus commented on lib/Doctrine/Migrations/Configuration/Connection/Loader/ArrayConnectionConfigurationLoader.php:
```diff
@@ -17,14 +17,18 @@
 
-    public function __construct(?string $filename)
+    /** @var ConnectionLoaderInterface */
+    private $fallback;
```
Ok 👍

alcaeus commented:
@goetas thanks for applying feedback, I've left comments on the open items.

We'll have to check out the Windows failures, but I'd have to find the time to do so. If you have a chance to take a look, I'd appreciate it.

Now for the million dollar question: are you planning to rewrite the commit history before merging? 😅

goetas committed:
filesystem sorting is not relevant in the finder

goetas committed:
migration plan execution is immutable

goetas committed:
warn to not use registerMigrationInstance

goetas committed:
regex finder accepts any class naming

goetas committed:
do not allow invalid config keys

goetas committed:
array loader is a hard dependency for the other loaders

goetas committed:
no invalid keys are allowed

goetas committed:
refer to https://github.com/Roave/BackwardCompatibilityCheck instead of having a long list

goetas commented:
@alcaeus I think that almost all the issues have been addressed (windows tests are running too).
If there are no other issues, I think this is ready to be merged. More changes can be done after it in the master branch.


1) Regarding the git history rewrite, I think that is pretty hard. I have two options:

    1) squash everything in a single commit (deletes a lot of history, but at least will be a single clean commit)
   2) leave commits as they are (brings in some mess, but still keeps the history of hat happend and how...)

   I personally would opt to keep the history as it is, unless someone has a better idea.

2) What has not been addressed yet are some exception-renamings, named constructors and more value-objects. I consider those change very valuable, but at the moment not worth spending the already my limited time on it. Would be very happy to see a pull request by someone else with the proposed changes.

3) Things to be handled in successive pull requests are:
    - What to do with the `down()` migration
    - How to gracefully upgrade to 3.0 the migrations table (see https://github.com/goetas/migrations/pull/1)
    - What to do with the sql instructions for the migration table in case of dry-run file dumps (see https://github.com/doctrine/migrations/pull/858#discussion_r327189524)

lcobucci commented:
@goetas I understand that rewriting the history might feel a bit overwhelming, even scary. However, if we can do a (soft) reset (`git reset upstream/master`) and adding changes interactively (`git add -p`), this process becomes much simpler and easy.
It's still a lot of work but it would greatly improve things and make this changeset more understandable (squashing everything isn't a viable option too since it'd turn things into a massive commit).

Don't forget we're here to help you on this too :+1:

goetas commented:
@lcobucci So if I understand you correctly, the idea is to do a soft reset and do few commits based on folders/files? 
In this way the commits will be still kinda messy but at least limited to some files.... In this way we still lose the real history, but the commits will be something as:

```
- changed console commands
- added metadata
- removed functional tests
- bla bla
- ...
```

is that what you mean?

greg0ire commented:
>  based on folders/files

`git add -p` allows you to commit part of files. Hopefully you would be able to group cs fixes/SA tools fixes together, and so on, rather than adding all the changes in a file.

goetas commented:
> git add -p allows you to commit part of files. Hopefully you would be able to group cs fixes/SA tools fixes together, and so on, rather than adding all the changes in a file.

sounds a hell of a job

greg0ire commented:
> sounds a hell of a job

It's a bit cumbersome with big PRs like this. I can try to do it for you if you're not feeling comfortable with this.

goetas committed:
classgenerator test

lcobucci commented:
@goetas just saw you pushed new things here. Did you finish this batch of work? I was going to reorganise the commits today but if things are still going to change it's better to wait (especially now that the build isn't green anymore).

goetas commented:
I was testing it locally and noticed that there was a bug. So I've added
the fix and tests. Hope it does not happen again.

On Tue, 22 Oct 2019, 22:00 Luís Cobucci, <notifications@github.com> wrote:

> @goetas <https://github.com/goetas> just saw you pushed new things here.
> Did you finish this batch of work? I was going to reorganise the commits
> today but if things are still going to change it's better to wait
> (especially now that the build isn't green anymore).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/doctrine/migrations/pull/858?email_source=notifications&email_token=AAF5UJZYORCOVN3RIJ64EG3QP5LVHA5CNFSM4IWZ62MKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB7AACA#issuecomment-545128456>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF5UJ33JT4W6L7BE3Y365LQP5LVHANCNFSM4IWZ62MA>
> .
>

goetas committed:
ensure there is always some metadata configuration set

goetas commented:
https://github.com/goetas/migrations/pull/2 is a Draft PR for making more consistent the down migrations when not implemented

lcobucci commented:
I'll get the bug fix applied, thanks

goetas commented:
Once more improvement proposed here https://github.com/goetas/migrations/pull/3
Makes the configuration and dependency injection container immutable. (frozen state)
This will avoid having other people hacking into the DI container or config object at runtime (avoiding inconsistent retrieval of services

goetas commented:
@alcaeus @lcobucci Any news on this? Can we merge it if there are no things to change? 
So I can continue with the other pull request I've prepared in the past days.

lcobucci commented:
@goetas sorry, I'm still working on organising this change set in a sensible way.

It's quite awesome that you had time to do more but I haven't had much. So, please let's not rush things.

slince commented:
@lcobucci  Hello, I think if there are not too many problems, we can merge first. After merging, we can make some adjustments and fixes based on the latest code.  The PR has been reviewed for too long.

greg0ire commented:
@slince Luis isn't working on the code itself, but on the commits, which can't be altered afterwards. Please be patient.

goetas commented:
A couple of other changes/fixes in https://github.com/goetas/migrations/pull/4 to be applied in the next stage

lcobucci commented:
@goetas as @alcaeus requested I've pushed my work [to a fork](https://github.com/lcobucci/doctrine-migrations/commits/v30) so that you both can see the (slow) progress.

You'll find notes in most of the commits, which are basically things that I'm finding and was going to use as the source to my review of this huge branch. There's still a lot to go, though.

I hope this helps in the visibility of things :+1:

alcaeus commented on lib/Doctrine/Migrations/Version/Version.php:
```diff
@@ -4,237 +4,26 @@
     }
 
-    public function getVersion() : string
+    public function __toString()
```
Hard to catch, but this removes the return type from the method. Was this on purpose?

From: https://github.com/lcobucci/doctrine-migrations/pull/1/commits/b4e07acfce8b6380504e67517d367583e69d420d.

goetas replied:
Will fix. The reason for it was a complain by php, or the method signature... but must be a mistake, looking at https://3v4l.org/DHAEs should work!

alcaeus commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -0,0 +1,141 @@
+        $this->schemaManager = $connection->getSchemaManager();
+        $this->platform      = $connection->getDatabasePlatform();
+
+        if ($configuration !== null && ! ($configuration instanceof TableMetadataStorageConfiguration)) {
```
We could change the type hint to `TableMetadataStorageConfiguration`, as widening it to `MetadataStorageConfiguration` later (once we support different configurations) is perfectly acceptable without a BC break.

lcobucci replied:
This is nothing but a side-effect of https://github.com/doctrine/migrations/pull/858#discussion_r344870590, though. Solving the root cause should allow us to remove this condition.

alcaeus commented on lib/Doctrine/Migrations/Metadata/Storage/MetadataStorageConfigration.php:
```diff
@@ -0,0 +1,9 @@
+
+namespace Doctrine\Migrations\Metadata\Storage;
+
+interface MetadataStorageConfigration
```
Apart from the typo (thanks @lcobucci), are there any methods we can/should add to the interface?

Copied from https://github.com/lcobucci/doctrine-migrations/pull/1/commits/1f59f350aae1d96ba10e7c3e645fa34994787fcc

goetas replied:
will fix

goetas replied:
Since each "metadata storage" might have complete different ways to store data, and the configurations for that storage are used only buy the storage, there are no shared configs that could be placed in this interface


This should answer to https://github.com/doctrine/migrations/pull/858#discussion_r345026244

alcaeus commented on phpcs.xml.dist:
```diff
@@ -47,17 +47,18 @@
         <exclude-pattern>lib/Doctrine/Migrations/Version/ExecutorInterface.php</exclude-pattern>
         <exclude-pattern>lib/Doctrine/Migrations/ParameterFormatterInterface.php</exclude-pattern>
-        <exclude-pattern>lib/Doctrine/Migrations/Version/VersionInterface.php</exclude-pattern>
+        <exclude-pattern>lib/Doctrine/Migrations/MigratorInterface.php</exclude-pattern>
```
Since @lcobucci brought it up in his review, I'd appreciate it if we can revisit interface names here.

goetas replied:
Ok, will do

alcaeus commented on phpcs.xml.dist:
```diff
@@ -47,17 +47,18 @@
         <exclude-pattern>tests/Doctrine/Migrations/Tests/Configuration/AbstractConfigurationTest.php</exclude-pattern>
         <exclude-pattern>tests/Doctrine/Migrations/Tests/Tools/Console/Command/AbstractCommandTest.php</exclude-pattern>
+        <exclude-pattern>tests/Doctrine/Migrations/Tests/Configuration/Loader/AbstractLoaderTest.php</exclude-pattern>
+        <exclude-pattern>lib/Doctrine/Migrations/Configuration/Loader/AbstractFileLoader.php</exclude-pattern>
```
I'd appreciate it if we could revisit the naming here to potentially drop the `Abstract` prefix.

goetas replied:
Ok, will do

alcaeus commented on tests/Doctrine/Migrations/Tests/TestLogger.php:
```diff
@@ -0,0 +1,57 @@
+use function strpos;
+use function strtr;
+
+class TestLogger extends AbstractLogger
```
Any reason why we can't use [`Psr\Log\Test \TestLogger`](https://github.com/php-fig/log/blob/446d54b4cb6bf489fc9d75f55843658e6f25d801/Psr/Log/Test/TestLogger.php#L57) here?

Copied from https://github.com/lcobucci/doctrine-migrations/pull/1/commits/37954e3c3790a53a819906e0ea5e22eddd3054ea

goetas replied:
Did not notice that. Will do!

goetas replied:
I checked better, the PSR test logger is not doing the string interpolation, that makes easier to test text messages.

alcaeus reviewed (changes requested):
I've added some review comments copied over from the PR @lcobucci linked.

@goetas can you take a look at the feedback please?

goetas commented:
I plan to continue working on this tomorrow and fix the comments you've left.

goetas committed:
rename AliasResolver into DefaultAliasResolver

goetas committed:
rename AbstractCommand into DoctrineCommand

goetas committed:
removed cs rules for outdated classes

goetas committed:
renamed ConfigurationHelperInterface into ConfigurationHelper

goetas committed:
renamed ConnectionLoaderInterface into ConnectionLoader

goetas committed:
removed cs rules for outdated classes

goetas committed:
renamed ParameterFormatter into InlineParameterFormatter

goetas committed:
renamed FileBuilderInterface into ConcatenationFileBuilder

goetas committed:
renamed ExecutorInterface into Executor

goetas committed:
renamed Migrator into DbalMigrator

goetas committed:
add typehint on version class

goetas committed:
fixed typo in MetadataStorageConfiguration class

goetas committed:
re sort imports after class renaming

goetas commented:
I managed to do already the fixes you asked (except of https://github.com/doctrine/migrations/pull/858#discussion_r347105103)

lcobucci commented on lib/Doctrine/Migrations/Generator/DefaultFileBuilder.php:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @internal
+ */
+final class DefaultFileBuilder implements FileBuilder
```
Using "default" as prefix isn't quite clear. Perhaps `ConcatenationFileBuilder` or something similar would be more explicit.

goetas replied:
You are right. Your suggestion is much better.
I'm also not happy with the other "default" named classes, but could not get better ideas...

greg0ire commented on lib/Doctrine/Migrations/DefaultParameterFormatter.php:
```diff
@@ -0,0 +1,97 @@
+ *
+ * @internal
+ */
+final class DefaultParameterFormatter implements ParameterFormatter
```
I read you were not happy with that prefix… may I suggest `InlineParameterFormatter` here? Since it produces one line with all the parameters, as opposed to a table? Should  the test class be renamed too?

goetas replied:
Good idea! thanks!

goetas commented:
Latest changes pushed!

OskarStark commented:
Congratulations 🍾🎉🎊🎈

Karimgasmi47 commented on lib/Doctrine/Migrations/Tools/Console/Command/DiffCommand.php:
```diff
@@ -100,12 +98,12 @@ public function execute(
         InputInterface $input,
```
Can we also change the visibility from "public" to "protected" on the Command class to match the declaration with Symfony/Console ?
If you agree, I may open a PR

goetas replied:
Sure! Did not know was protected

ostrolucky commented:
I must say decision to use FQCN instead of version number is causing us issues. We have deployed branch with doctrine/migrations 3.x to our staging environment. It migrated the table. Now other branches with older version of doctrine/migrations fail because they don't recognize versions as executed anymore. For the next time, there should be a BC layer in currently maintained branch which recognizes both formats.
